### PR TITLE
Linecuts2d

### LINKED ISSUE #180: Additional plotting style
Hello,

I was working with some data in plottr lately and missed a way to show a 2D dataset as the different linecuts. I know you can plot the linecuts one by one, however for some data it is more illustrative to show all the linecuts in the same plotwindow.  I wrote a simple function and implementation of these kind of plots, see below:

```
def plot_as_linecuts(ax: Axes, x: np.ndarray, y: np.ndarray,
                     z: np.ndarray, cmap, **kw):
    
    im = "linecuts"
    sm = plt.cm.ScalarMappable(cmap=cmap, norm=plt.Normalize(vmin=np.amin(y), vmax=np.amax(y)))
    
    if len(x.shape) == 1:
        ax.plot(x,z,color = sm.to_rgba((y[0,0])),**kw)
    else:
        for i in range(x.shape[1]):
            ax.plot(x[:,i],z[:,i],color = sm.to_rgba((y[0,i])),**kw)
    
    return im, sm
```

It is very simple and can probably made a bit smarter by adding masked arrays to show continuous progress, now it just plots a new linecut after a step on the slow axis.

I was wondering if others are interested in this extension as an addition to the current plottypes and if so would it be possible to add something similar to this to a future release of plottr. If this is something I should keep in my own fork please let me know then I can close this topic.

### PR DESCRIPTION
**Pull request by MdKquantum**
As discussed earlier in the similar named issue. Here is my implementation of a 2D meshgrid plotted as individual linecuts. The linecuts are taken along the fast axis. 

I don't have an icon at the moment so the button is just a text label. 

If there are any questions or suggestions let me know.

### CODE REVIEW

File: plottr/plot/mpl.py [hunk 1]
```diff
@@ -254,14 +258,22 @@ def colorplot2d(ax: Axes, x: np.ndarray, y: np.ndarray, z: np.ndarray,
...
+        im, sm = plot_as_linecuts(ax,x,y,z,cmap=cmap, **kw)
 
     if im is None:
         return
 
-    cax = attachColorBar(ax, im)
-    ax.set_xlabel(axLabels[0])
-    ax.set_ylabel(axLabels[1])
-    cax.set_ylabel(axLabels[2])
+    if type(im) == Line2D:
```
Comment (astafan8) [inline plottr/plot/mpl.py]: the more robust/correct way of checking the type in python ;)
```suggestion
    if isinstance(im, Line2D):
```

File: plottr/plot/mpl.py [hunk 2]
```diff
@@ -292,6 +304,40 @@ def ppcolormesh_from_meshgrid(ax: Axes, x: np.ndarray, y: np.ndarray,
...
+    if isinstance(y, np.ma.MaskedArray):
+        y = y.filled(np.nan)
+    if isinstance(z, np.ma.MaskedArray):
+        z = z.filled(np.nan)
+    
+    if len(x.shape) == 1:
+        line, =ax.plot(x,z,color = sm.to_rgba((y[0,0])),**kw)
+    else:
+        for i in range(x.shape[1]):
+            line, = ax.plot(x[:,i],z[:,i],color = sm.to_rgba((y[0,i])), **kw)
```
Comment (astafan8) [inline plottr/plot/mpl.py]: PEP8 ;)
```suggestion
        line, = ax.plot(x, z, color=sm.to_rgba((y[0, 0])), **kw)
    else:
        for i in range(x.shape[1]):
            line, = ax.plot(x[:, i], z[:, i], color=sm.to_rgba((y[0, i])), **kw)
```

File: plottr/plot/mpl.py [hunk 3]
```diff
@@ -292,6 +304,40 @@ def ppcolormesh_from_meshgrid(ax: Axes, x: np.ndarray, y: np.ndarray,
...
+    """Plot a 2D array as the individual linecuts
+    Input are the corresponding arrays from a 2D ``MeshgridDataDict``.
+
+    Will attempt to fix missing points in the coordinates.
+
+    :param ax: axes to plot the colormesh into.
+    :param x: x component of the meshgrid coordinates
+    :param y: y component of the meshgrid coordinates
+    :param z: data values
+    :returns: the image returned by `pcolormesh`.
```
Comment (astafan8) [inline plottr/plot/mpl.py]: is this true? the function returns the line and also the ScalarMappable. also the return type of this function is missing :)

File: plottr/plot/mpl.py [hunk 4]
```diff
@@ -292,6 +304,40 @@ def ppcolormesh_from_meshgrid(ax: Axes, x: np.ndarray, y: np.ndarray,
...
+
+    Keywords are passed on to `plot`.
```
Comment (astafan8) [inline plottr/plot/mpl.py]: clarification
```suggestion
    Keywords are passed on to `Axes.plot`.
```

File: plottr/plot/mpl.py (hunk 3 above)
Comment (MdKquantum) [reply plottr/plot/mpl.py]: I will check the added comments, I added those quickly without properly looking at them

**Comment by MdKquantum** [comment]
Is there anything else that should be done, or are we awaiting the refactoring of the plotting system?

**Comment by wpfff** [comment]
updating the refactored plotting code with this feature would be quick i think. 
If @astafan8, @FarBo, and @jenshnielsen think that the refactored plotting is close to merge-ready then we could just wait and do it there.
(waiting for an answer before i do it, though :))

**Comment by FarBo** [comment]
@wpfff 
I think it is close to be merged. I am looking forward to test the refactoring PR after it is done.

**Comment by jenshnielsen** [comment]
@MdKquantum the big refactor has now landed so we can look at merging this again. Unfortunately these files have moved a bit. I think the most of your code will now need to go into mpl/plotting.py

**Comment by MdKquantum** [comment]
I think I just pushed a version of this plotstyle in here with the plot refactor taken in to account. 

To do this elegantly we might have to look into how to minimize the amount of if statements in plotting.py and autoplot.py. I checked locally and everything works how I have implemented it, but it might not be the most elegant.

On a side note @wpfff I had the pyqt5 5.15.4 installed and couldn't launch the whole inspectr any more. After a pyqt downgrade everything worked again. I have the traceback for pyqt 5.14.4 saved if you are intrested.

**Comment by wpfff** [comment]
thanks @MdKquantum -- @jenshnielsen or @astafan8 -- could one of you look into the inspectr issue with latest qt? we're not using it here since we don't use the qcodes dataset, so it'd take me a bit to investigate.

**Comment by jenshnielsen** [comment]
@wpfff  sure @MdKquantum could you open a new issue with the traceback you are seeing

**Comment by jenshnielsen** [comment]
I have not my self seen any issues with the latest version of pyqt5

Running
```
PyQt5                         5.15.4
PyQt5-Qt5                     5.15.2
PyQt5-sip                     12.9.0
PyQt5-stubs                   5.15.2.0
pyqtgraph                     0.12.1
PyQtWebEngine                 5.15.4
PyQtWebEngine-Qt5             5.15.2
```

**Comment by MdKquantum** [comment]
I think I cleared the error on the previous check now, still not quite sure how to read those, in case there is any more issues let me know.

**Comment by MdKquantum** [comment]
What needs to be done to make this release ready? 

I don't quite understand why it keeps failing the checks, while this version works on my test PC.

**Comment by jenshnielsen** [comment]
@MdKquantum We enforce that all the code in plottr is type annotated: Specifically these are the errors that you get

```
plottr/__init__.py:102:36: error: Argument 1 to "module_from_spec" has incompatible type "Optional[ModuleSpec]"; expected "ModuleSpec"
plottr/__init__.py:104:31: error: Item "None" of "Optional[ModuleSpec]" has no attribute "loader"
plottr/__init__.py:105:13: error: Item "None" of "Optional[ModuleSpec]" has no attribute "loader"
plottr/plot/mpl/plotting.py:222:1: error: Function is missing a type annotation for one or more arguments
```

File: plottr/plot/mpl/plotting.py [hunk 5]
```diff
@@ -206,3 +218,29 @@ def plotImage(ax: Axes, x: np.ndarray, y: np.ndarray,
     im = ax.imshow(z.T, aspect='auto', origin='lower',
                    extent=extent, **kw)
     return im
+    
+def plot_as_linecuts(ax: Axes, x: np.ndarray, y: np.ndarray,
+                     z: np.ndarray, cmap, **kw) -> Tuple[str,ScalarMappable]:
```
Comment (jenshnielsen) [inline plottr/plot/mpl/plotting.py]: Here cmap and **kw are not typed. 
cmap should be of type `matplotlib.colors.Colormap` e.g. here `colors.Colormap` should be correct.

**kw can just be typed as Any

Comment (jenshnielsen) [reply plottr/plot/mpl/plotting.py]: the other failures are not related to your pr

Comment (jenshnielsen) [reply plottr/plot/mpl/plotting.py]: They should be fixed by https://github.com/toolsforexperiments/plottr/pull/207

**Comment by MdKquantum** [comment]
@wpfff Can we merge this after the branch has been updated? Or do further checks have to be passed?

**Comment by wpfff** [comment]
code wise this looks OK, but i have maybe a functionality question :)
normally when you see these kinds of plots they're typically 'waterfall' plots, where the individual traces are offset vertically. here they're just on top of each other which created in a simple test i just ran a bit of a useless mess. would it maybe be good to add an offset?

**Comment by MdKquantum** [comment]
an offset is useful, the challenge will then be to determine how to get this offset suitable for everything. I guess we could do an offset as the percentage of the maximum amplitude?

**Comment by wpfff** [comment]
yes, something like that would probably make sense. the important thing for this feature to be universally useful is that most data looks clear, i think.
i could also imagine that there's a user option for that. this could maybe be included in the GUI as an option.

**Comment by astafan8** [comment]
hey @MdKquantum , this PR is very interesting and useful but you had conversation with @wpfff above that hasn't resulted in any commits yet - do you intend to continue working on this PR ? :)

**Comment by MdKquantum** [comment]
I'm currently in the middle of writing a publication which makes my time a bit limited. I'm hoping to get some more free time soon too finish this. The Slider and/or other changes on the GUI would require some more work from my side. Sorry for the delay

**Comment by wpfff** [comment]
nothing happened here in a while, and this branch is very outdated. closing for now, can re-open if needed.

**Pull request closed without merging**